### PR TITLE
expose metrics via EventEmitter

### PR DESCRIPTION
**Pull request by dt-rush**
Change-type: patch
Signed-off-by: Nick Payne <nickp@balena.io>

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,24 +1,34 @@
...
 export class MaxSizeExceededError extends PromiseQueueError {}
 export class TimeoutError extends PromiseQueueError {}
 
+const durationSince = (t0: ReturnType<typeof process.hrtime>): number => {
+	const diff = process.hrtime(t0);
+	return diff[0] * 1000 + diff[1] / 1e6;
+};
+
 export class PromiseQueue {
+	public name: string;
```
Comment (Page-) [inline src/index.ts]: Why is the name added? It seems like something that would be handled/used purely in dependents code - it's not actually used anywhere in this module

File: src/index.ts [hunk 2]
```diff
@@ -79,16 +96,25 @@ export class PromiseQueue {
 						reject(e);
 						return;
 					}
+					this.metrics.emit('queueTime', durationSince(enqueueTime));
+					this.metrics.emit('dequeue');
+					const serviceStartTime = process.hrtime();
 					const result = await fn();
+					this.metrics.emit('serviceTime', durationSince(serviceStartTime));
```
Comment (Page-) [inline src/index.ts]: These never get emitted in the case of a failure, so it's probably better to put them in a finally

File: src/index.ts (hunk 1 above)
Comment (dt-rush) [reply src/index.ts]: True, I can see in the branch on the API that there are "pool" objects where this ought to live.

File: src/index.ts (hunk 2 above)
Comment (dt-rush) [reply src/index.ts]: Ah, good catch

File: src/index.ts (hunk 1 above)
Comment (dt-rush) [reply src/index.ts]: Removed all references to name

File: src/index.ts (hunk 2 above)
Comment (dt-rush) [reply src/index.ts]: Now fixed

File: src/index.ts [hunk 3]
```diff
@@ -70,25 +82,37 @@ export class PromiseQueue {
 			}
 
 			let timeout: ReturnType<typeof setTimeout> | undefined;
+			let enqueueTime: ReturnType<typeof process.hrtime>;
+			let serviceStartTime: ReturnType<typeof process.hrtime>;
```
Comment (Page-) [inline src/index.ts]: Can you declare this inside `wrappedFn` please, there's no need to scope it outside

File: src/index.ts [hunk 4]
```diff
@@ -70,25 +82,37 @@ export class PromiseQueue {
...
 					reject(e);
+				} finally {
+					this.metrics.emit('serviceTime', durationSince(serviceStartTime));
+					this.metrics.emit('latency', durationSince(enqueueTime));
+					this.metrics.emit('completion');
 				}
 			};
 
 			this.queue.push(wrappedFn);
+			enqueueTime = process.hrtime();
```
Comment (Page-) [inline src/index.ts]: Can you declare/assign this in one go to a const please, which also removes the need to explicitly type it

**Review by Page- — COMMENTED** [review]
This should be a minor bump since it introduces new functionality

File: src/index.ts (hunk 3 above)
Comment (dt-rush) [reply src/index.ts]: ack

File: src/index.ts (hunk 4 above)
Comment (dt-rush) [reply src/index.ts]: ack

File: src/index.ts [hunk 5]
```diff
@@ -42,6 +49,7 @@ export class PromiseQueue {
 		this.maxAge = maxAge;
 		this.order = order;
 		this.next = order === 'lifo' ? 'pop' : 'shift';
+		this.metrics = new EventEmitter();
```
Comment (Page-) [inline src/index.ts]: Can you put this above as
```ts
public metrics = new EventEmitter();
```
please

File: src/index.ts [hunk 6]
```diff
@@ -55,9 +63,15 @@ export class PromiseQueue {
 		}
 	}
 	public add<T>(fn: () => T | PromiseLike<T>): Promise<T> {
+		this.metrics.emit('arrival');
+		this.metrics.emit('queueLength', this.queue.length);
+		this.metrics.emit('inFlight', this.inFlight);
+		const enqueueTime = process.hrtime();
```
Comment (Page-) [inline src/index.ts]: Can you move this inside the `new Promise` call to limit it to just the scope it's needed

**Review by Page- — COMMENTED** [review]
Can you delete the package-lock.json please, it doesn't make sense for modules imo as downstream users will ignore it

File: src/index.ts (hunk 5 above)
Comment (dt-rush) [reply src/index.ts]: good point

File: src/index.ts (hunk 6 above)
Comment (dt-rush) [reply src/index.ts]: done

**Review by Page- — APPROVED** [review]

**Pull request merged**